Trac 2849, Pretty property names moved to the properties themselves

alf239 opened the pull request:
- DB changes are done
- migration scripts are ready
- Display names are supported in:
  - Experiment page
  - Tabs over charts
  - Gene table
  - Gene table search dropdown
  - Experiment design
  - Search results
  - Experiment popup in heatmap
  - Experiment index
  - Advanced search
  - EF dropdown
  - Search
    - EF in conditions lookup dropdown
    - Search results (list view)
  - Gene page's EF headers
- `LoaderDAO` generates accessions itself (ugly, but still)

Broken or missing functionality (comes soon after): 
- Display name editing via admin interface
- EF tabs in experiment list on gene page

Commit:
Trac 2849 - Display Name for properties and values - data structures

Commit:
Trac 2849 - Display Name for properties and values - PropertyDAO instead of AtlasProperties

Commit:
Trac 2849 - Display Name for properties and values - accession instead of curatedName

Commit:
Trac 2849 - Display Name for properties and values - curated names are in DB now, no need to keep them around in properties

Commit:
Merge remote branch 'origin/develop' into trac2849

Commit:
Trac 2849 - Display Name for properties and values - curated names from curators' instance

Commit:
Trac 2849 - Display Name for properties and values - ExperimentDesign page migrated to *.displayNames

Commit:
Trac 2849 - Display Name for properties and values - isolating JSP changes

Commit:
Trac 2849 - Display Name for properties and values - returning bold assay accessions

Commit:
Trac 2849 - Display Name for properties and values - prettifying factors without displayName set

Commit:
Trac 2849 - Display Name for properties and values - dropping explicit definitions for factors which can be prettified automatically

Commit:
Trac 2849 - Display Name for properties and values - dropping explicit definitions for factors which can be prettified automatically

Commit:
Trac 2849 - Display Name for properties and values - fixing the migrations' conflict

Commit:
Trac 2849 - Display Name for properties and values - Tabs over charts; Experiment popup in heatmap

Commit:
Trac 2849 - Display Name for properties and values - Experiment index

Commit:
Merge remote branch 'origin/develop' into trac2849

Commit:
Trac 2849 - Display Name for properties and values - Gene with one EF only — header

Commit:
Trac 2849 - Display Name for properties and values - The fake property storage

Commit:
Trac 2849 - Display Name for properties and values - @Deprecated the old way

alf239 committed:
Dropping unused method

alf239 committed:
Spelling

alf239 committed:
Dropping unused property

Commit:
Merge remote branch 'origin/develop' into trac2849

Commit:
Keeping things simple: parameter only used for logging does more harm than good

Commit:
Garbage collection

Commit:
Trac 2849: Sanitized property accessions — Basic escaping on load

Commit:
Merge remote branch 'gxa/develop' into trac2849

Conflicts:
	atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Property.java

Commit:
Trac 2849 Removing Capitalization In Ef Names

Commit:
Trac 2849 Encapsulated property accession constraints, fixed tests

Commit:
Trac 2849 Encapsulated further

Commit:
nicer diff

Commit:
Trac 2849 name/accession should never be null

Commit:
Trac 2849 definition/value should never be null

Commit:
Merge remote branch 'gxa/develop' into trac2849

Conflicts:
	atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Assay.java:
```diff
@@ -244,6 +247,18 @@ public class Assay {
         }
     }
 
+    public Collection<PropertyValue> getEffectiveValues(Property property) {
```
Could you add javadoc on how you define an 'effective' value

alf239 replied:
here you are

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/AssayProperty.java:
```diff
@@ -60,15 +60,7 @@ public final class AssayProperty {
-        this.terms = new ArrayList<OntologyTerm>(efoTerms);
-    }
-
-    public AssayProperty(Long id, Assay assay, PropertyValue pv, List<OntologyTerm> efoTerms) {
```
Why does Assay have a constructor with id in it, but AssayProperty doesn't. Is id in Assay just used for junit testing?

alf239 replied:
Historical reasons mostly; I'll take a look

alf239 replied:
Thanks for spotting—dropped the id

alf239 committed:
javadoc on `getEffectiveValues(Property property)`

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Experiment.java:
```diff
@@ -330,4 +330,15 @@ public class Experiment {
         return null;
     }
+
+    public SortedSet<Property> getProperties() {
```
Not sure I understand - this method is called in e.g. AtlasExperiment.getFactorValuesForEF().keys(). Going by the parent method name, shouldn't keys return just EF names, rather than EF and SC names?

alf239 replied:
It's a bug actually, listed in outstanding tasks in http://bar.ebi.ac.uk:8080/trac/ticket/2849

Will review.

alf239 replied:
Addressed in 4693210

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/PropertyValue.java:
```diff
@@ -4,25 +4,34 @@ import org.hibernate.annotations.Cache;
 
     public PropertyValue(Long id, Property definition, String value) {
+        if (definition == null)
+            throw new NullPointerException("Definition must be provided");
```
Looking at calls to this constructor, definition should never be null. Should we not throw unexpected exception instead here?

alf239 replied:
It doesn't change anything, and should never happen, anyway. So I'm happy to change it to anything at all, as long as it's unchecked. Don't think NPE is bad or unsuitable, though.

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Sample.java:
```diff
@@ -235,17 +209,13 @@ public class Sample {
-    }
-
-    public boolean hasProperty(final PropertyValue propertyValue) {
+    private boolean hasNoProperty(final PropertyValue propertyValue) {
```
Personally I find hasNoProperty method name confusing for two reasons - a. it sounds almost the same as another method: hasNoProperties() and at first sight sounds like both methods do the same b. Having No in the name confuses me more than than !hasProperty(). Could we not revert to the previous method definition?

alf239 replied:
I don't think `hasNoProperties()` can create a problem here: one method has a parameter, another hasn't. On the inversion, there are different schools of though, so in order to save time I've rewritten the code.

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Sample.java:
```diff
@@ -258,12 +228,29 @@ public class Sample {
         }
     }
+
+    public Collection<PropertyValue> getEffectiveValues(Property property) {
```
javadoc on the meaning of 'effective' in the new scheme of things

alf239 replied:
Renamed

alf239 replied:
Javadoc added, too.

rpetry commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/SampleProperty.java:
```diff
@@ -22,21 +22,17 @@
```
Same question as for AssayProperty - why does Sample have id in its constructor, but SampleProperty doesn't?

alf239 replied:
One I've cleaned up (unused constructor), and another was used somewhere. It wasn't a cleanup change, so I've only dropped what's obviously unused.

alf239 replied:
Thanks for spotting—dropped the id

Commit:
Trac 2849 Dropping confusing method

Commit:
Merge remote branch 'gxa/trac2849' into trac2849

alf239 commented on atlas-updates/src/main/resources/db/migrations/201110171551-property-display-names.sql:
```diff
@@ -0,0 +1,18 @@
+update a2_property set displayname='Clinical info' where name='clinical_information';
+update a2_property set displayname='Compound treatment' where name='compound';
+update a2_property set displayname='Genotype' where name='genotype';
```
Sources were:
1. `Curated.properties`
2. `select * from a2_config_property where name like 'factor.curatedname.%'`

alf239 replied:
By the way, leaving `update a2_property set displayname='Genotype' where name='genotype';` was a mistake: I was trying not to set display names where `StringUtil.prettify()` would do the same.

No need to remove, though—it's harmless.

Commit:
Trac 2849 Javadoc, method names, rationale behind the multiple-value approach

rpetry commented on atlas-web/src/main/java/ae3/service/structuredquery/AtlasEfvService.java:
```diff
@@ -84,14 +87,16 @@ public class AtlasEfvService implements AutoCompleter, IndexBuilderEventHandler,
     }
 
-    public Set<String> getOptionsFactors() {
+    public Set<Property> getOptionsFactors() {
```
Would you mind changing getOptionsFactors() to getOptionsIgnoredFactors(). The intended function is somewhat lost in the name otherwise.

alf239 replied:
I would actually: it returns all the factors _except_ ignored, so  `getOptionsIgnoredFactors()` would be even more confusing. I would be fine with `getOptionFactorsWithoutTheFactorsWeIgnore()`, but that looks too verbose and Monty Python-ish.

rpetry replied:
Ah sorry - my mistake - you're right - it's fine as it is

rpetry commented on atlas-web/src/main/java/uk/ac/ebi/gxa/properties/EfHeadersStorage.java:
```diff
@@ -38,10 +40,10 @@ public class EfHeadersStorage implements Storage {
     private final static String PREFIX = "factor.curatedname.";
     private static final int SUFFIXPOS = PREFIX.length();
```
no longer used

alf239 replied:
bloody survivors! removed.

Commit:
Trac 2849 Unused constant

rpetry commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/GenePageExperiment.java:
```diff
@@ -8,6 +9,7 @@ import uk.ac.ebi.microarray.atlas.model.Experiment;
 
 import java.util.Set;
 
+@JsonSerialize
```
What does this annotation give us? And if needed, why doesn't GenePageGene have it too?

alf239 replied:
It shows that the class will be serialized with Jackson to avoid confusion with usage search. As of `GenePageGene`, didn't change it—not in scope.

rpetry commented on atlas-web/src/main/webapp/WEB-INF/jsp/experimentpage/experiment.jsp:
```diff
@@ -311,9 +312,9 @@
                                 <option value="">All factor values</option>
                                 <c:forEach var="EF" items="${exp.experimentFactors}">
-                                    <optgroup label="${f:escapeXml(atlasProperties.curatedEfs[EF])}">
+                                    <optgroup label="${f:escapeXml(EF.displayName)}">
```
Do we need to `f:escapeXml(EF.displayname)`? I didn't need to in [experiment-index.jsp](https://github.com/gxa/gxa/blob/develop/atlas-web/src/main/webapp/WEB-INF/jsp/experimentpage/experiment-index.jsp).

alf239 replied:
We should always escape the values we put on the page unless we know they are sanitized. See also [OWASP Top 10 2010-A2-Cross-Site Scripting (XSS)](https://www.owasp.org/index.php/Top_10_2010-A2)

alf239 replied:
By the way, here's the code from experiment-index.jsp:

``` jsp
                    <c:forEach var="factor" items="${experiment.experimentFactors}">
                        ${f:escapeXml(atlasProperties.curatedEfs[factor])}<br/>
                    </c:forEach>
```

rpetry commented on atlas-web/src/main/webapp/WEB-INF/jsp/includes/atlas-searchform.jsp:
```diff
@@ -198,8 +198,8 @@
-                        <c:forEach var="i" items="${atlasQueryService.experimentalFactorOptions}">
-                            <option value="${f:escapeXml(i)}">${f:escapeXml(atlasProperties.curatedEfs[i])}</option>
+                        <c:forEach var="factor" items="${atlasQueryService.experimentalFactorOptions}">
+                            <option value="${factor.name}">${f:escapeXml(factor.displayName)}</option>
```
Do we need to `f:escapeXml`?

alf239 replied:
Imagine a pretty name like "Dose for 0 < t <= 5"—we'd better replace `<` with `&lt;`. 

I'm changing it to `<c:out value="" />` where appropriate (`c:out` escapes XML by default), but this one hasn't got there.

Commit:
Trac 2849 Unused IDs in Samples and Assays; general cleanup

Conflicts:

	atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Assay.java
	atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Sample.java

alf239 committed:
Dropped unused line—just in case

Commit:
Merge remote branch 'gxa/develop' into trac2849

Conflicts:
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/GeneViewController.java
	atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp

Commit:
Apparently dropped by mistake

Commit:
Making difference between experiment factors (used almost everywhere) and experiment properties (i.e., EFs + SCs) used only in experiment design

alf239 commented:
On `breast cancer cell line`, it's a data problem; we must have properties sanitised in DB or else index will be useless. We may of course sanitize properties as we read those from DB—but that would mean we don't trust our own data, which would be an extremely weak position.
